### PR TITLE
Feature processqueue

### PR DESCRIPTION
**Pull request by roynilsson**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets | #37 

This feature adds a queue system where processes that should be run later can be added. A key benefit is to be able to create several processes and then run them one at a time. Could for example be imports sent to us from external systems. We dont want 4 different heavy processes updating (potentially the same) products at the same time. 

A new model, QueueItem, represents the queued process. It fills a similar role as the Executable model. It holds which type of process should be run, settings/params for that process and status in the queue. There is also a string field to name the queue. This allows us to have multiple queues. Eg our order export process goes in one queue and supplier product imports goes in another. That way we can still export orders to our ERP system even if we are importing products at the same time.

There is a new interface QueueAwareProcessInterface that extends the ProcessInterface. This adds two methods runFromQueue and canRun. runFromQueue starts the process just like the run method from the ProcessInterface. The difference is that implementations of runFromQueue must start the process in the foreground. Most, if not all, implementations of run that I've come across start the process in the background. We must start the process in the foreground so that we can update status of the queue item after it completes. If we start processes in the background we can not keep track of when they complete and thus dont know when it is time to start the next.

canRun is called before starting the process to see if it should run now or not. This allows every implementation of a queue aware process to provide its own logic for when and how queued items should be started.

There is a console command to run items off the queue. This could/should be scheduled as a cron job running every minute. The command gets queue items with the 'queued' status, calls canRun on them and then runFromQueue if they could run. Since runFromQueue should start the process in the foreground this command will be blocked until the process completes. The command can then set success or fail status on the queue item.

The originial idea was to have the queue runner as a Pimcore maintenance task rather than a command. However we do not want potentially long running processes blocking the maintenance job.

There is a gui to see the queue in admin but there is no way to add new items to the queue or modify existing items.

### CODE REVIEW

**Comment by roynilsson** [comment]
I opened the issue for this nearly a year ago. Got tied down by other projects but finally came back to it now.

Comments and critique is most welcome

**Comment by dpfaffenbauer** [comment]
I am quite busy at the moment, but I will review your PR as soon as I find time, ok?

**Comment by roynilsson** [comment]
No worries, take your time

File: src/ProcessManagerBundle/Maintenance/QueueTask.php [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+/**
+ * Process Manager.
+ *
+ * LICENSE
+ *
+ * This source file is subject to the GNU General Public License version 3 (GPLv3)
+ * For the full copyright and license information, please view the LICENSE.md and gpl-3.0.txt
+ * files that are distributed with this source code.
+ *
+ * @copyright  Copyright (c) 2018 Jakub Płaskonka (jplaskonka@divante.pl)
```
Comment (kubaplas) [inline src/ProcessManagerBundle/Maintenance/QueueTask.php]: @roynilsson I am not the author of this file :)

Comment (dpfaffenbauer) [reply src/ProcessManagerBundle/Maintenance/QueueTask.php]: :D, Copyright is always me, even if someone else created the file... at least it should be me...

Comment (kubaplas) [reply src/ProcessManagerBundle/Maintenance/QueueTask.php]: I guess copyright to you, but `@author` can be set to @roynilsson ?

Comment (roynilsson) [reply src/ProcessManagerBundle/Maintenance/QueueTask.php]: You're such a great developer @kubaplas that you write code without knowing it! Or maybe I made a copy-paste oops. I will fix that

**Comment by dpfaffenbauer** [comment]
@roynilsson I finally had time to review your PR. And it is amazing :). I just updated master to support Pimcore X and Pimcore X only. Can you add docs on how to use Queue Items?

**Comment by dpfaffenbauer** [comment]
@roynilsson could you please provide docs?

**Review by dpfaffenbauer — APPROVED** [review]

**Pull request closed without merging**